components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -2643,15 +2643,25 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
             throw new ResponseError(ErrorCodes.BAD_REQUEST, "organization ID must be a valid UUID");
         }
 
-        const checkAgainstDB = async (): Promise<{ team: Team; members: TeamMemberInfo[] }> => {
-            // We deliberately wrap the entiry check in try-catch, because we're using Promise.all, which rejects if any of the promises reject.
-            const team = await this.teamDB.findTeamById(teamId);
-            if (!team) {
-                // We return Permission Denied because we don't want to leak the existence, or not of the Organization.
-                throw new ResponseError(ErrorCodes.PERMISSION_DENIED, `No access to Organization ID: ${teamId}`);
-            }
+        const team = await this.teamDB.findTeamById(teamId);
+        if (!team) {
+            // We return Permission Denied because we don't want to leak the existence, or not of the Organization.
+            throw new ResponseError(ErrorCodes.PERMISSION_DENIED, `No access to Organization ID: ${teamId}`);
+        }
 
-            const members = await this.teamDB.findMembersByTeam(team.id);
+        const members = await this.teamDB.findMembersByTeam(team.id);
+
+        await this.guardTeamOperationInternal(team, members, op, fineGrainedOp);
+        return { team, members };
+    }
+
+    private async guardTeamOperationInternal(
```

Comment:
this method could have a bit of a clearer contract such as `hasOrgPermission` and then return boolean for the normal case and throw an error in exceptional cases.